src/Analyzers/Core/Analyzers/RemoveUnnecessarySuppressions/AbstractRemoveUnnecessaryPragmaSuppressionsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,558 @@
+                // without updating the below code.
+
+                var (assembly, compilerAnalyzerTypeName) = GetCompilerDiagnosticAnalyzerInfo();
+                var compilerAnalyzerType = assembly.GetType(compilerAnalyzerTypeName)!;
+                var methodInfo = compilerAnalyzerType.GetMethod("GetSupportedErrorCodes", BindingFlags.Instance | BindingFlags.NonPublic)!;
+                var compilerAnalyzerInstance = Activator.CreateInstance(compilerAnalyzerType);
+                var supportedCodes = methodInfo.Invoke(compilerAnalyzerInstance, Array.Empty<object>()) as IEnumerable<int>;
+                return supportedCodes.ToImmutableHashSet();
+            }
+            catch (Exception ex)
+            {
+                Debug.Fail(ex.Message);
+                return ImmutableHashSet<int>.Empty;
+            }
+        }
+
+        public sealed override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        protected sealed override void InitializeWorker(AnalysisContext context)
+        {
+            // We do not register any normal analyzer actions as we need 'CompilationWithAnalyzers'
+            // context to analyze unused suppressions using reported compiler and analyzer diagnostics.
+            // Instead, the analyzer defines a special 'AnalyzeAsync' method that should be invoked
+            // by the host with CompilationWithAnalyzers input to compute unused suppression diagnostics.
+        }
+
+        public async Task AnalyzeAsync(
+            SemanticModel semanticModel,
+            TextSpan? span,
+            CompilationWithAnalyzers compilationWithAnalyzers,
+            Func<DiagnosticAnalyzer, ImmutableArray<DiagnosticDescriptor>> getSupportedDiagnostics,
+            Func<DiagnosticAnalyzer, bool> getIsCompilationEndAnalyzer,
+            Action<Diagnostic> reportDiagnostic,
+            CancellationToken cancellationToken)
+        {
+            // We need compilation with suppressed diagnostics for this feature.
+            if (!compilationWithAnalyzers.Compilation.Options.ReportSuppressedDiagnostics)
+            {
+                return;
+            }
+
+            var tree = semanticModel.SyntaxTree;
+
+            // Bail out if analyzer is suppressed on this file or project.
+            // NOTE: Normally, we would not require this check in the analyzer as the analyzer driver has this optimization.
+            // However, this is a special analyzer that is directly invoked by the analysis host (IDE), so we do this check here.
+            if (tree.DiagnosticOptions.TryGetValue(IDEDiagnosticIds.RemoveUnnecessarySuppressionDiagnosticId, out var severity) ||
+                compilationWithAnalyzers.Compilation.Options.SpecificDiagnosticOptions.TryGetValue(IDEDiagnosticIds.RemoveUnnecessarySuppressionDiagnosticId, out severity))
```

Comment:
do we have tests for these sorts of things.

Reply:
Yes, we have test for this.

Reply:
See `TestDoNotRemoveDiagnosticSuppressionsForSuppressedAnalyzer`